HIPE 0161: generic token proposal (#160)
hyperledger/indy-hipe: text/0161-generic-token/README.md

# 0161: Generic Tokens

> - Author: [Lynn Bendixsen](lynn@indicio.tech)

Reviewer: ```suggestion
- Authors: [Brent Zundel](brent.zundel@evernym.com), [Lynn Bendixsen](lynn@indicio.tech)
```

### To provide a payment mechanism for the masses
Many Indy ledger users desire a network designed with a utility payment mechanism to allow for on-ledger transfer of value. Building this payment plugin opportunity directly into the main codebase will fill that need in the simplest, and most direct way.

## Tutorial
[tutorial]: #tutorial

### Introduction
The generic token can be defined and used in any manner an Indy ledger implementer would like to. Consensus around valid token transactions occurs with the same pool of validator nodes as every other transaction. Fees that accompany transactions are processed atomically with those transactions, so they fail or succeed together.

The infrastructure for value transfer on a ledger consists of a payments sub-ledger (much like the domain and pool sub-ledgers) that records payment transactions. Setting fees writes a fee schedule to the existing config sub-ledger and setting Auth_Rules allows for those fees to be applied to any desired transactions. There is also a cache of unspent transaction outputs, or _UTXOs_.

> 

Author: ```suggestion

This document proposes optional payment and fees plugins that enable any Indy ledger to support a token. 
```

   - For example: if the address pay:tok:8675309 was sent tokens in transactions with sequence numbers 129, 455, and 1090, and none of the tokens has been spent, then the key would be 1:pay:tok:8675309 and the value would be 129:455:1090.

The UTXO cache is updated whenever a payment transaction is written to the ledger.

### Transaction Fees
The primary purpose of transaction fees is to reduce spam and to introduce a financial disincentive for DDoS and similar attacks. A ledger operator can offer open access by setting the fees for ledger transactions high enough to discourage abuse of the network while at the same time keeping fees as low as possible.

The fees are set and maintained by the ledger operator. The transactions for which fees are collected are the transactions that require the most work from validator nodes, namely requests to write to the ledger.

These transactions are configurable, but typically are the following:
- NYM
- ATTRIB
- SCHEMA
- CRED_DEF
- REVOC_REG_DEF
- REVOC_REG_ENTRY
- XFER_PUBLIC

Transactions with fees may be added or removed from this list.

### Incorporation with Hyperledger Indy

> Hyperledger Indy provides the code base for Indy ledgers. This code base may be extended with plugins. Ledger plugins add new transactions and sub-ledgers to the transactions and sub-ledgers already defined by Hyperledger Indy.

Reviewer: ```suggestion
Hyperledger Indy provides the code base for Indy ledgers. This code base may be extended with plugins. Ledger plugins add new transactions and sub-ledgers to the transactions and sub-ledgers already defined by Hyperledger Indy. This document introduces plugins for generic token transactions, which may be incorporated into any Indy ledger.
```

Two ledger plugins are proposed in this document:
- A Generic Token plugin, which adds a payments sub-ledger and the following transactions:
   - **MINT_PUBLIC**
   - **XFER_PUBLIC**
   - **GET_UTXO**
- A Fees plugin, which adds the following transactions to the payments sub-ledger mentioned above:
   - **GET_FEES**
   - **SET_FEES**
   - **FEES**

Note: While the Generic Token plugin and the accompanying payment sub-ledger may be used without the Fees plugin, the Fees plugin depends on the Generic Token plugin.

#### Indy-SDK Payments API Plugin
The Indy-SDK payments API comes with a default payment handler plugin called
[LibNullPay](https://github.com/hyperledger/indy-sdk/blob/master/libnullpay/README.md). LibNullPay allows for a "null" token to be used with the payments API. The "null" token transactions are dummy transactions. They are not stored on a ledger and do not initiate consensus.

The payments API allows for other payment handler plugins to be initialized. An Indy-SDK payments API compatible payment handler is proposed in this document:

> - LibTokToken, which handles generic token payment functionality through the Indy-SDK payments API for the Token and Fees ledger plugins and produces properly formatted and signed transaction requests for each of the new transactions listed above. LibTokToken also parses the transaction responses from the new ledger plugins through the Indy-SDK payments API.

Reviewer: LibSovToken was a terrible name. I highly recommend not copying too closely here.
```suggestion
- LibIndyToken, which handles generic token payment functionality through the Indy-SDK payments API for the Token and Fees ledger plugins and produces properly formatted and signed transaction requests for each of the new transactions listed above. LibIndyToken also parses the transaction responses from the new ledger plugins through the Indy-SDK payments API.
```

## Reference
[reference]: #reference

Documentation for the proper format of transaction requests and transaction responses for the generic token plugin, specifically for the **MINT_PUBLIC**, **XFER_PUBLIC**, and **GET_UTXO** transactions can be found here: [Ledger Token Transactions.](TODO: add link)

Documentation for the proper format of transaction requests and transaction responses for the Fees plugin, specifically for the **SET_FEES**, **GET_FEES**, and **FEES** transactions can be found here: [Ledger Fee Transactions.](TODO: add link)

Documentation for the Indy-SDK payments API and how to use it may be found here: [Indy SDK Payments API.](https://github.com/hyperledger/indy-sdk/tree/master/docs/design/004-payment-interface)

> An example of the structure of the inputs and outputs of the LibTokToken payment handler plugin for the Indy-SDK payments API are documented here: [Token specific data structures.](https://github.com/sovrin-foundation/libsovtoken/blob/master/doc/data_structures.md) (from LibTokTokens predecessor LibSovToken)

Reviewer: ```suggestion
An example of the structure of the inputs and outputs of the LibIndyToken payment handler plugin for the Indy-SDK payments API are documented here: [Token specific data structures.](https://github.com/sovrin-foundation/libsovtoken/blob/master/doc/data_structures.md) (from LibIndyToken's predecessor LibSovToken)
```

## Drawbacks
[drawbacks]: #drawbacks

### Undesired Speculation
The intention of the generic token is to enable a more public way to submit transaction requests or provide for an alternative payment method without compromising the security and performance of the ledger validator nodes. However, the introduction of the token may be incorrectly viewed as an invitation for speculators to purchase generic tokens, even though they have no intention of making use of the utility provided by the generic token.

This speculative action may cause the value of a generic token to fluctuate, requiring a quick response from the ledger operator in adjusting the fee schedule for ledger transactions so that fees remain reasonable.

### Additional Costs

> Moving from an Endorser model, where trusted entities can submit write transaction to the ledger, to a Token model, where submitting a write transaction requires the payment of fees denominated in generic tokens, will likely introduce new costs to those who are currently trusted entities.

Reviewer: ```suggestion
Moving from an Endorser model, where trusted entities can submit write transaction to the ledger, to a Token model, where submitting a write transaction requires the payment of fees denominated in generic tokens, will likely introduce new costs to those who are currently trusted entities, unless the Auth rules are formulated to explicitly support both models.
```

- the **FEE** transaction should not be written if the primary transaction fails.

Recording the **FEE** transaction on the payment sub-ledger and the primary transaction on another sub-ledger may require that the two sub-ledgers be synchronized in some way. This may also cause difficulties during catchup (when a validator node updates its state to be current with the rest of the validator pool). This might no longer be a drawback due to the introduction of the audit ledger.

If the generic token begins to be commonly used for other payments besides fees, there will be additional load on the validator nodes as they come to consensus on the **XFER_PUBLIC** transactions.

## Rationale and alternatives
[alternatives]: #alternatives

This design for a generic token described here is simple. It is based on the UTXO model for transactions found in [bitcoin](https://bitcoin.org/en/bitcoin-paper), which simplifies validity checking for payment transactions.

If a generic token is not introduced, every network desiring token based payment transfers must write their own plug-in and build their own branding.

## Prior art
[prior-art]: #prior-art

> This proposal was inspired by the Sovrin Token and its associated [SIP](https://github.com/sovrin-foundation/sovrin-sip/tree/master/text/5003-sovrin-tokens). 

Reviewer: ```suggestion
This proposal is a generic form of the Sovrin Token, made suitable for use by any Indy ledger, and the text is primarily taken from the associated [SIP](https://github.com/sovrin-foundation/sovrin-sip/tree/master/text/5003-sovrin-tokens). 
```

## Unresolved questions
[unresolved]: #unresolved-questions

There are a number of questions that remain unanswered, some of which may be

> outside of the scope of this effort:

Reviewer: some additional questions it may be worthwhile to consider:
```suggestion
outside of the scope of this effort:
- Should fees paid to the ledger remain there (i.e., fees represent spent tokens), or should they be collected (i.e., fees represent tokens paid to an entity)?
- If the fees are to be collected by some entity, how is it determined what entity should collect them?
- What mechanism should be used to collect fees paid to the ledger? An automatic or a manual one?
- How might fees paid to the ledger be automatically distributed among participating validator nodes?
```

The drawback of this protection mechanism is that it makes the ledger less accessible to anyone outside of the trusted circle. If an Indy ledger is truly to be a global public utility, the ability to request that a transaction be written to the ledger should be available to anyone. This is the primary motivation for introducing a token as a payment mechanism for transaction requests.

If submitting a transaction request to a validator did not require a party to be trusted, but instead required a small fee, this would provide an economic disincentive to abuse the ledger while allowing anyone to submit transaction requests.

### To provide a payment mechanism for the masses
Many Indy ledger users desire a network designed with a utility payment mechanism to allow for on-ledger transfer of value. Building this payment plugin opportunity directly into the main codebase will fill that need in the simplest, and most direct way.

> 

Author: ```suggestion
### A common API
Providing a common API for all Indy tokens will enable better wallet support. A wallet that supports a single Indy token could support any Indy token.
```

The transaction types handled by the payment ledger are:
- **MINT_PUBLIC** - Creates generic tokens as a transaction on the payment ledger. The minting of tokens typically requires the signatures of multiple trustees, but is configurable.
- **XFER_PUBLIC** - Transfers generic tokens from a set of input payment addresses to a set of output payment addresses. This transaction is recorded on the payment ledger.
- **GET_UTXO** - Queries the payment ledger for unspent transactions held by a payment address.
- **SET_FEES** - Sets the fee schedule for ledger transactions, including transactions recorded on other sub-ledgers. This transaction is recorded on the config ledger. The setting of fees typically requires the signatures of multiple trustees, but is configurable.
- **GET_FEES** - Queries the config ledger to retrieve the current fee schedule.
- **FEES** - Attaches a fee payment, according to the Auth_Rules in conjunction with the fee schedule, to transactions requests. The fee payment is recorded on the payment ledger. This is never an independent transaction, but always attached to the primary transaction request for which fees are being paid.

> 

Reviewer: ```suggestion

- **COLLECT_FEES** - Collects previously uncollected fees accumulated on the ledger, according
to the Auth_Rules. Each fee payment that has been recorded to the ledger, but not yet included
in a **COLLECT_FEES** transaction will be collected and deposited into the single specified
payment address. The **COLLECT_FEES** transaction records the transaction number of the first
fee transaction included in the current deposit and always collects all subsequent fees for
transactions up to the time of this **COLLECT_FEES** transaction.
```

- the primary transaction should not be processed if the **FEE** transaction fails,
- the **FEE** transaction should not be written if the primary transaction fails.

Recording the **FEE** transaction on the payment sub-ledger and the primary transaction on another sub-ledger may require that the two sub-ledgers be synchronized in some way. This may also cause difficulties during catchup (when a validator node updates its state to be current with the rest of the validator pool). This might no longer be a drawback due to the introduction of the audit ledger.

If the generic token begins to be commonly used for other payments besides fees, there will be additional load on the validator nodes as they come to consensus on the **XFER_PUBLIC** transactions.

## Rationale and alternatives
[alternatives]: #alternatives

This design for a generic token described here is simple. It is based on the UTXO model for transactions found in [bitcoin](https://bitcoin.org/en/bitcoin-paper), which simplifies validity checking for payment transactions.

If a generic token is not introduced, every network desiring token based payment transfers must write their own plug-in and build their own branding.

> 

Reviewer: ```suggestion

As an alternative to the COLLECT_FEES transaction, it may be possible to
relegate the collection of fees to the FEE transaction itself, in conjunction
with an addition to the config ledger which specifies the payment address
to which fees should be sent. The assignment of this payment address
could be specified as its own transaction, as an addition to the SET_FEES
transaction, or as an addition to the MINT_PUBLIC transaction (e.g., reclaim
paid fees to the original minting address). The benefits of this alternative
over the COLLECT_FEES transaction include improved auditability, but the
drawbacks include potential negative performance impacts and a possibly
unmanageable number of small payments accumulated in the destination
payment address, so this alternative was deemed unviable at this time.

Another alternative to the COLLECT_FEES transaction is to never collect fees.
The safeguards against ledger attacks remain, but the tokens spent as fees
would be unrecoverable. It is possible to mimic this scenario by constituting
the ledger Auth_Rules such that the COLLECT_FEES transaction may not be called.

```

Transactions with fees may be added or removed from this list.

### Incorporation with Hyperledger Indy
Hyperledger Indy provides the code base for Indy ledgers. This code base may be extended with plugins. Ledger plugins add new transactions and sub-ledgers to the transactions and sub-ledgers already defined by Hyperledger Indy. This document introduces plugins for generic token transactions, which may be incorporated into any Indy ledger.

#### Ledger Plugins
The process for adding plugins to Hyperledger Indy is explained in [Hyperledger's indy-plenum repository](https://github.com/hyperledger/indy-plenum/blob/master/docs/source/plugins.md).

Plugins allow for new transactions to be added to an instantiation of Indy, without requiring changes to the core codebase. New sub-ledgers may also be defined using these plugins. A sub-ledger is where the new transactions may be stored.

Two ledger plugins are proposed in this document:
- A Generic Token plugin, which adds a payments sub-ledger and the following transactions:
   - **MINT_PUBLIC**
   - **XFER_PUBLIC**
   - **GET_UTXO**
- A Fees plugin, which adds the following transactions to the payments sub-ledger mentioned above:

>    - **GET_FEES**
>    - **SET_FEES**
>    - **FEES**

Reviewer: ```suggestion
   - **GET_FEES**
   - **SET_FEES**
   - **FEES**
   - **COLLECT_FEES**
   - **UNCOLLECTED_FEES**
```

The Indy-SDK payments API comes with a default payment handler plugin called
[LibNullPay](https://github.com/hyperledger/indy-sdk/blob/master/libnullpay/README.md). LibNullPay allows for a "null" token to be used with the payments API. The "null" token transactions are dummy transactions. They are not stored on a ledger and do not initiate consensus.

The payments API allows for other payment handler plugins to be initialized. An Indy-SDK payments API compatible payment handler is proposed in this document:
- LibToken, which handles generic token payment functionality through the Indy-SDK payments API for the Token and Fees ledger plugins and produces properly formatted and signed transaction requests for each of the new transactions listed above. LibToken also parses the transaction responses from the new ledger plugins through the Indy-SDK payments API.

## Reference
[reference]: #reference

Documentation for the proper format of transaction requests and transaction responses for the generic token plugin, specifically for the **MINT_PUBLIC**, **XFER_PUBLIC**, and **GET_UTXO** transactions can be found here: [Ledger Token Transactions.](TODO: add link)

> Documentation for the proper format of transaction requests and transaction responses for the Fees plugin, specifically for the **SET_FEES**, **GET_FEES**, and **FEES** transactions can be found here: [Ledger Fee Transactions.](TODO: add link)

Reviewer: ```suggestion
Documentation for the proper format of transaction requests and transaction
responses for the Fees plugin, specifically for the **SET_FEES**, **GET_FEES**,
**FEES**, **COLLECT_FEES**, and **UNCOLLECTED_FEES** transactions can be
found here: [Ledger Fee Transactions.](TODO: add link)
```
